### PR TITLE
Add a post on the build system

### PR DESCRIPTION
**Pull request by larsbergstrom**
r? @metajack 

Wrote this up from some notes I was already preparing on Servo's build system for @alexcrichton anyway :-)

[Review on Reviewable](https://reviewable.io/reviews/servo/blog.servo.org/20)

### CODE REVIEW

File: _posts/2015-07-21-build.md [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+
+The rest of Servo is a significant amount of code (~150k lines of Rust; ~250k if you include autogenerated DOM bindings), but follows the standard conventions of Cargo and Rust as far as producing crates. For the many crates within the Servo repo, we simply have a `Cargo.toml` file next to a `lib.rs` that defines the module structure. When we break them out into a separate GitHub repository, though, we follow the convention of a toplevel `Cargo.toml` file with a `src` directory that holds all of the Rust code.
+
+![Servo's dependency graph](https://raw.githubusercontent.com/maxsnew/cargo-dot/master/etc/servo.png "Servo dependency graph")
+
+## Things we'd like to fix in the future
+
+It would be great if there was a single Cargo.toml file and it was at the toplevel of the Servo repo. It's confusing to people familiar with Rust projects, who go looking for a Cargo.toml file and can't find them.
```
Comment (SimonSapin) [inline _posts/2015-07-21-build.md]: Link to https://github.com/servo/servo/issues/5973 ?

File: _posts/2015-07-21-build.md [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+
+Servo has several testing tools that can be executed via mach.
+
+* `mach tidy` will verify that there are no trivial syntactic errors in source files. It checks for valid license headers in each file, no tab characters, no trailing whitespaces, etc.
+
+* `mach test-ref` will run the Servo-specific reference tests. These tests render a pair of web pages that implement the same final layout using different CSS features to images. If the images are not pixel-identically, the test fails.
+
+* `mach test-wpt` runs the cross-browser W3C Web Platform Tests, which primarily test DOM features.
+
+* `mach test-css` runs the cross-browser CSS WG refernce tests, which are a version of the reference tests that are intended to work across many browsers.
```
Comment (jdm) [inline _posts/2015-07-21-build.md]: s/refernce/reference/

File: _posts/2015-07-21-build.md [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+
+## Mach
+
+[Mach](https://developer.mozilla.org/en-US/docs/Mozilla/Developer_guide/mach) is a python driver program that provides a frontend to Servo's build that both reduces the number of steps to build and integrates our different testing tools into a single frontend harness. Similar to its purpose in the Firefox build, we use it to centralize and simplify the number of commands that a developer has to perform.
+
+### mach bootstrap
+
+The steps that mach will handle before issuing a normal `cargo build` command are:
+* Downloading the correct versions of the cargo and rustc tools. Servo uses most of the unstable features in Rust, which change pretty frequently. We also test the edges of feature compatibility and so are the first ones to notice many changes that did not at first seem as if they would break anyone. Further, we build a custom version of the tools that additionally supports cross-compilation targeting Android (and ARM in the near future). A random local install of the Rust toolchain is pretty unlikely to work with Servo.
```
Comment (Manishearth) [inline _posts/2015-07-21-build.md]: > Servo uses most of the unstable features in Rust

I'd say this has changed. We use some of the _most_ unstable features of Rust, but actually our unstable usage is pretty low other than plugins and some APIs here and there.

> local install of the Rust toolchain is pretty unlikely to work with Servo

I wonder if we split plugins into lints and expansion (expansion plugins are generally more stable) and made it possible to turn lints off, we might get a decently large error box around the rustc version.

File: _posts/2015-07-21-build.md [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+
+* `mach tidy` will verify that there are no trivial syntactic errors in source files. It checks for valid license headers in each file, no tab characters, no trailing whitespaces, etc.
+
+* `mach test-ref` will run the Servo-specific reference tests. These tests render a pair of web pages that implement the same final layout using different CSS features to images. If the images are not pixel-identically, the test fails.
+
+* `mach test-wpt` runs the cross-browser W3C Web Platform Tests, which primarily test DOM features.
+
+* `mach test-css` runs the cross-browser CSS WG reference tests, which are a version of the reference tests that are intended to work across many browsers.
+
+* `mach test-unit` runs the Rust unit tests embedded in Servo crates. We do not have many of these, except for basic tests of per-crate functionality.
```
Comment (Manishearth) [inline _posts/2015-07-21-build.md]: Might be worth mentioning that we don't have many of these because wpt and css cover most bases.

**Comment by larsbergstrom** [comment]
r? @SimonSapin @Ms2ger 

I think I've addressed everybody's comments - should I squash & land this?

**Comment by Manishearth** [comment]
lgtm, r=me

**Pull request merged**